cdap-proto/src/main/java/co/cask/cdap/proto/id/EntityId.java

```diff
@@ -142,6 +142,51 @@ public MetadataEntity toMetadataEntity() {
     throw new UnsupportedOperationException("Metadata is not supported");
   }
 
+  @Nullable
+  public static <T extends NamespacedEntityId> T fromMetadataEntity(MetadataEntity metadataEntity) {
+    EntityType targetType = findTargetType(metadataEntity);
+    if (targetType == null) {
+      return  null;
+    }
+    if ((targetType == EntityType.APPLICATION || targetType == EntityType.PROGRAM) &&
+      metadataEntity.getValue(MetadataEntity.VERSION) == null) {
+      if (targetType == EntityType.APPLICATION) {
+        metadataEntity = metadataEntity.append(MetadataEntity.VERSION, ApplicationId.DEFAULT_VERSION);
+      } else {
+        metadataEntity = MetadataEntity.ofNamespace(metadataEntity.getValue(MetadataEntity.NAMESPACE))
+          .append(MetadataEntity.APPLICATION, metadataEntity.getValue(MetadataEntity.APPLICATION))
+          .append(MetadataEntity.VERSION, ApplicationId.DEFAULT_VERSION)
+          .append(MetadataEntity.TYPE, metadataEntity.getValue(MetadataEntity.TYPE))
+          .append(MetadataEntity.PROGRAM, metadataEntity.getValue(MetadataEntity.PROGRAM));
+      }
+    }
+    T entityId;
+    try {
+      entityId  = targetType.fromIdParts(metadataEntity::getValues);
+    } catch (IllegalArgumentException e) {
+      entityId = null;
+    }
+    return entityId;
+  }
+
+  @Nullable
+  private static EntityType findTargetType(MetadataEntity metadataEntity) {
+    List<MetadataEntity.KeyValue> keyValues = metadataEntity.getKeyValues();
+
+    EntityType entityType = null;
+    // we want to identify the target type from child walking up to parent
+    int curIndex = keyValues.size() - 1;
+    while (curIndex >= 0) {
```

Comment:
this is ugly. Why not have a static Map<String, EntityType> and look it up there?